Add B+/Bd lifetime ratio as observable

MJKirk opened the pull request:
Includes contributions from (sbar b)(cbar c), (sbar b)(cbar u) and (dbar b)(cbar u) operators.

Requires merging my wilson PR (https://github.com/wilson-eft/wilson/pull/116) so the sbcu and dbcu sectors are available

MJKirk committed:
Add B+/Bd lifetime ratio observable

Currently includes BSM contributions from all operators in the dbcu and sbcu
sectors, as well as all (dbar b)(cbar c) operators.

peterstangl reviewed (changes requested):
I think we should require wilson v2.5 to make sure the new WET sectors are available:

https://github.com/flav-io/flavio/blob/4eca6c8883edb8bc80586254176339a9142b0264/setup.py#L35-L36

peterstangl commented on flavio/physics/bdecays/lifetime_ratio.py:
```diff
@@ -0,0 +1,301 @@
+        ((3*rho**0.5*(me["5"] - 2*me["6"]))/4,(9*rho**0.5*(me["5"] - 2*me["6"]))/4,(3*me["6"])/2,(9*me["6"])/2,0,0,0,0,0,0,0,0,0,0,(-3*rho**0.5*(me["1p"] - 2*me["2p"]))/8,(-9*rho**0.5*(me["1p"] - 2*me["2p"]))/8,(-((1 + 2*rho)*me["1p"]) + 2*(2 + rho)*me["2p"])/8,(-3*(me["1p"] + 2*rho*me["1p"] - 2*(2 + rho)*me["2p"]))/8,(0.5 + rho)*me["1p"] - (-4 + rho)*me["2p"],(3*(me["1p"] + 2*rho*me["1p"] - 2*(- ...
+        (-(rho**0.5*(me["5"] + 2*(me["6"] + 3*me["7"] + 6*me["8"]))),-3*rho**0.5*(me["5"] + 2*me["6"]),6*(me["6"] + 6*me["8"]),18*me["6"],0,0,0,0,0,0,0,0,0,0,(rho**0.5*(me["1p"] + 2*(me["2p"] + 3*me["3p"] + 6*me["4p"])))/2,(3*rho**0.5*(me["1p"] + 2*me["2p"]))/2,(me["1p"] + 2*rho*me["1p"] - 2*(-4 + rho)*me["2p"] + 6*(me["3p"] + 2*rho*me["3p"] + 8*me["4p"] - 2*rho*me["4p"]))/6,(0.5 + rho)*me["1p"]  ...
+        (-3*rho**0.5*(me["5"] + 2*me["6"]),-9*rho**0.5*(me["5"] + 2*me["6"]),18*me["6"],54*me["6"],0,0,0,0,0,0,0,0,0,0,(3*rho**0.5*(me["1p"] + 2*me["2p"]))/2,(9*rho**0.5*(me["1p"] + 2*me["2p"]))/2,(0.5 + rho)*me["1p"] - (-4 + rho)*me["2p"],(3*(me["1p"] + 2*rho*me["1p"] - 2*(-4 + rho)*me["2p"]))/2,-2*(1 + 2*rho)*me["1p"] + 4*(14 + rho)*me["2p"],-6*(me["1p"] + 2*rho*me["1p"] - 2*(14 + rho)*me["2p"] ...
+    ))
```
Maybe this could be made a bit more readable by creating the array from a list instead of a tuple, and by adding line breaks and indentation similar to the following example:
```python
    A_WE_cu = np.array([
        [
            (-((2 + rho)*me["1"]) + 2*(me["2"] + 2*rho*me["2"] - 3*(2 + rho)*me["3"] + 6*(me["4"] + 2*rho*me["4"])))/6,
            -((2 + rho)*me["1"])/2 + me["2"] + 2*rho*me["2"],
            -(rho**0.5*(me["2"] + 6*me["4"])),
            -3*rho**0.5*me["2"],
            0,0,0,0,0,0,0,0,0,0,
            ((2 + rho)*me["5p"] - 2*(me["6p"] + 2*rho*me["6p"] - 3*(2 + rho)*me["7p"] + 6*(me["8p"] + 2*rho*me["8p"])))/12,
            ((2 + rho)*me["5p"] - 2*(me["6p"] + 2*rho*me["6p"]))/4,
            (rho**0.5*(me["5p"] - 2*(me["6p"] - 3*me["7p"] + 6*me["8p"])))/4,
            (3*rho**0.5*(me["5p"] - 2*me["6p"]))/4,
            -(rho**0.5*(me["5p"] + 2*(me["6p"] + 3*me["7p"] + 6*me["8p"]))),
            -3*rho**0.5*(me["5p"] + 2*me["6p"]),
        ],
        [
            -((2 + rho)*me["1"])/2 + me["2"] + 2*rho*me["2"],
            (-3*(2 + rho)*me["1"])/2 + 3*(1 + 2*rho)*me["2"],
            -3*rho**0.5*me["2"],
            -9*rho**0.5*me["2"],
            0,0,0,0,0,0,0,0,0,0,
            ((2 + rho)*me["5p"] - 2*(me["6p"] + 2*rho*me["6p"]))/4,
            (3*((2 + rho)*me["5p"] - 2*(me["6p"] + 2*rho*me["6p"])))/4,
            (3*rho**0.5*(me["5p"] - 2*me["6p"]))/4,
            (9*rho**0.5*(me["5p"] - 2*me["6p"]))/4,
            -3*rho**0.5*(me["5p"] + 2*me["6p"]),
            -9*rho**0.5*(me["5p"] + 2*me["6p"]),
        ],
        [
            -(rho**0.5*(me["2"] + 6*me["4"])),
            -3*rho**0.5*me["2"],
            2*(me["2"] + 6*me["4"]),
            6*me["2"],
            0,0,0,0,0,0,0,0,0,0,
            (rho**0.5*(me["6p"] + 6*me["8p"]))/2,
            (3*rho**0.5*me["6p"])/2,
            (me["6p"] + 6*me["8p"])/2,
            (3*me["6p"])/2,
            6*(me["6p"] + 6*me["8p"]),
            18*me["6p"],
        ],
        ...
```

MJKirk replied:
Yes, that's probably nicer. I also considered whether it might be better to move these matrices off into a separate python file to keep this main file a bit cleaner, what do you think? (Also thinking a bit of the future if I/someone implements that contributions from many sectors, this could easily get huge)

DavidMStraub replied:
+1 for formatting, I would just format this entire file with `black`.

As for encapsulation, whether this thing is put into a separate module or not is secondary I would say, but it definitely should be a separate *function* (and have separate unit tests if possible...).

Something I would personally do (but up to you if you agree or not) is to use a `dataclass` instead of a `dict` for the lifetime matrix elements and the bag parameters. Although setting 18 parameters on it explicitly might seem less elegant than the for loop, it is more eplicit and easier for the type checker to understand than a dict with string keys.

peterstangl commented on flavio/physics/bdecays/lifetime_ratio.py:
```diff
@@ -0,0 +1,301 @@
+        ((3*rho**0.5*(me["5"] - 2*me["6"]))/4, (9*rho**0.5*(me["5"] - 2*me["6"]))/4, (1.5 - 3*rho)*me["6"], (4.5 - 9*rho)*me["6"], (-3*rho**0.5*me["5"])/8, (-9*rho**0.5*me["5"])/8, (-3*rho*me["5"])/4, (-9*rho*me["5"])/4, 3*rho*(me["5"] - 4*me["6"]), 9*rho*(me["5"] - 4*me["6"]), (3*rho**0.5*me["1p"])/4, (9*rho**0.5*me["1p"])/4, -3*rho*me["2p"], -9*rho*me["2p"], (-3*rho**0.5*(me["1p"] - 2*me["2p"]) ...
+        (-(rho**0.5*(me["5"] + 2*(me["6"] + 3*me["7"] + 6*me["8"]))), -3*rho**0.5*(me["5"] + 2*me["6"]), -6*(-1 + 2*rho)*(me["6"] + 6*me["8"]), -18*(-1 + 2*rho)*me["6"], (rho**0.5*(me["5"] - 4*me["6"] + 6*(me["7"] - 4*me["8"])))/2, (3*rho**0.5*(me["5"] - 4*me["6"]))/2, rho*(me["5"] - 4*me["6"] + 6*(me["7"] - 4*me["8"])), 3*rho*(me["5"] - 4*me["6"]), -4*rho*(me["5"] + 8*me["6"] + 6*(me["7"] + 8*me ...
+        (-3*rho**0.5*(me["5"] + 2*me["6"]), -9*rho**0.5*(me["5"] + 2*me["6"]), -18*(-1 + 2*rho)*me["6"], -54*(-1 + 2*rho)*me["6"], (3*rho**0.5*(me["5"] - 4*me["6"]))/2, (9*rho**0.5*(me["5"] - 4*me["6"]))/2, 3*rho*(me["5"] - 4*me["6"]), 9*rho*(me["5"] - 4*me["6"]), -12*rho*(me["5"] + 8*me["6"]), -36*rho*(me["5"] + 8*me["6"]), -3*rho**0.5*(me["1p"] - 4*me["2p"]), -9*rho**0.5*(me["1p"] - 4*me["2p"]) ...
+    ))
```
Same comment as above concerning readability.

peterstangl commented on flavio/physics/bdecays/lifetime_ratio.py:
```diff
@@ -0,0 +1,301 @@
+        (0,0,0,0,0,0,0,0,0,0,(-3*(me["5"] - 2*me["6"]))/4,(-me["5"] + 2*(me["6"] - 3*me["7"] + 6*me["8"]))/4,(-3*rho**0.5*me["6"])/2,-(rho**0.5*(me["6"] + 6*me["8"]))/2,(3*rho**0.5*(me["1"] - 2*me["2"]))/8,(rho**0.5*(me["1"] - 2*(me["2"] - 3*me["3"] + 6*me["4"])))/8,((2 + rho)*me["1"] - 2*(me["2"] + 2*rho*me["2"]))/8,((2 + rho)*me["1"] - 2*(me["2"] + 2*rho*me["2"] - 3*(2 + rho)*me["3"] + 6*(me["4 ...
+        (0,0,0,0,0,0,0,0,0,0,3*(me["5"] - 2*(me["6"] - 3*me["7"] + 6*me["8"])),9*(me["5"] - 2*me["6"]),-2*rho**0.5*(me["5"] - me["6"] + 6*me["7"] - 6*me["8"]),-6*rho**0.5*(me["5"] - me["6"]),-(rho**0.5*(me["1"] + 2*(me["2"] + 3*me["3"] + 6*me["4"])))/2,(-3*rho**0.5*(me["1"] + 2*me["2"]))/2,((-4 + rho)*me["1"] - 2*(me["2"] + 2*rho*me["2"] - 3*(-4 + rho)*me["3"] + 6*(me["4"] + 2*rho*me["4"])))/6,(( ...
+        (0,0,0,0,0,0,0,0,0,0,9*(me["5"] - 2*me["6"]),3*(me["5"] - 2*(me["6"] - 3*me["7"] + 6*me["8"])),-6*rho**0.5*(me["5"] - me["6"]),-2*rho**0.5*(me["5"] - me["6"] + 6*me["7"] - 6*me["8"]),(-3*rho**0.5*(me["1"] + 2*me["2"]))/2,-(rho**0.5*(me["1"] + 2*(me["2"] + 3*me["3"] + 6*me["4"])))/2,((-4 + rho)*me["1"])/2 - (1 + 2*rho)*me["2"],((-4 + rho)*me["1"] - 2*(me["2"] + 2*rho*me["2"] - 3*(-4 + rho) ...
+    ))
```
Same comment as above concerning readability.

peterstangl commented on flavio/physics/bdecays/lifetime_ratio.py:
```diff
@@ -0,0 +1,301 @@
+    ratio_SM = 1 + 0.059 * B1qtilde + 0.005 * B2qtilde - 0.674 * B3qtilde + 0.160 * B4qtilde \
+                 - 0.025 * deltaqq1tilde + 0.002 * deltaqq2tilde + 0.591 * deltaqq3tilde - 0.152 * deltaqq4tilde \
+                 - 0.007
+    return ratio_SM
```
Why is this phenomenological formula used here instead of the full expression? If possible, I think the full expression should always be used to allow for changes in the input parameters.

MJKirk replied:
Short answer: Because that's what I have access to.
Some justification: If you look at fig 5 in [2208.02643](https://arxiv.org/abs/2208.02643) you see that the SM uncertainty is ~75% coming from the bag parameters central values, and most of the rest is from $\mu_0$, which is the RG running of those values from the low scale where they are determined up to the B scale. So as long as you don't want to vary the other input parameters far outside their current uncertainties, this formula is valid.
Certainly it would be better to have the full formula, but I'd have to put it togther myself or ask Alex (or Maria Laura probably) if he has a notebook with everything in

peterstangl commented on flavio/physics/bdecays/test_lifetime_ratio.py:
```diff
@@ -0,0 +1,106 @@
+
+        wc = flavio.WilsonCoefficients()
+        wc.set_initial({"CVRLt_bdcc": -0.15, "CSLL_bdcc": 0.2}, scale=4.5)
+        self.assertAlmostEqual(flavio.np_prediction("tau_B+/tau_Bd", wc), 1.085435224)
```
Could it make sense to also test the size of the uncertainties in addition to the central values?

MJKirk replied:
Yes, good idea

MJKirk committed:
Increase required wilson version

In preparation for lifetime ratio observable. The newer wilson version
is needed since the sbcu and dbcu sectors were added in v2.5.

MJKirk commented on flavio/physics/bdecays/test_lifetime_ratio.py:
```diff
@@ -0,0 +1,210 @@
+        )
+        np.testing.assert_allclose(lifetime_ratio.A_WE_cc(0.1, me), A_WE_cc, atol=1e-3)
+
+    def test_A_PI_cd(self):
```
Ditto

MJKirk commented on flavio/physics/bdecays/lifetime_ratio.py:
```diff
@@ -0,0 +1,1577 @@
+
+
+def run_lifetime_bag_parameters(par, scale):
+    if scale < config["RGE thresholds"]["mc"] or scale > 4.5:
```
Add some delta around the limits of the running

MJKirk committed:
Improve tests of BSM functions in SM limit

MJKirk committed:
Add comments to lifetime ratio code

MJKirk committed:
Replace error with warning
